File: virtualfriends.cpp
#include <iostream>
#include <bits/stdc++.h>
#include <map>
using namespace std;

unordered_map<string, string> root;
unordered_map<string, vector<string> > sets;

void unite(string x, string y){
	string xparent = root[x];
	string yparent = root[y];
	if(xparent == yparent) return;
	if(sets[xparent].size() >= sets[yparent].size()){
		for(vector<string>::iterator it = sets[yparent].begin(); it != sets[yparent].end(); ++it){
			sets[xparent].push_back(*it);
			root[*it] = xparent;
		}
		sets[yparent].clear();
		sets[yparent].push_back(yparent);
	}
	else{
		for(vector<string>::iterator it = sets[xparent].begin(); it != sets[xparent].end(); ++it){
			sets[yparent].push_back(*it);
			root[*it] = yparent;
		}
		sets[xparent].clear();
		sets[xparent].push_back(xparent);
	}
	return;
}

string find(string s){
	if(root[s] == s){
		return s;
	}
	root[s] = find(root[s]);
	return root[s];
}

void addToSet(string s){
	vector<string> set = {s};
	root[s] = s;
	sets[s] = set;
}

int main(){
	ios::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);
	
	int n, m;
	string a, b;
	cin >> n;
	while(n--){
		cin >> m;
		while(m--){
			cin >> a >> b;
			if(root.find(a) == root.end()) addToSet(a);
			if(root.find(b) == root.end()) addToSet(b);
			unite(a, b);
			cout << sets[find(a)].size() << "\n";
		}
		root.clear();
		sets.clear();
	}
	return 0;
}

File: virtualfriends.py
#Virtual friends using disjoint set datastructure. Runtime 0.39s according to kattis.
from sys import stdin

class DisjointSet:
    def __init__(self):
        self.root = {}
        self.sets = {}

    def addSet(self, x):
        self.root[x] = x
        self.sets[x] = [x]

    def find(self, x):
        if self.root[x] == x:
            return x
        self.root[x] = self.find(self.root[x])
        return self.root[x]

    def unite(self, x, y):
        if self.root[x] == self.root[y]:
            return
        xparent = self.find(x)
        yparent = self.find(y)
        if len(self.sets[xparent]) >= len(self.sets[yparent]):
            for item in self.sets[yparent]:
                self.sets[xparent].append(item)
                self.root[item] = xparent
            self.sets[yparent] = [yparent]
        else:
            for item in self.sets[xparent]:
                self.sets[yparent].append(item)
                self.root[item] = yparent
            self.sets[xparent] = [xparent]

def main():
    cases = int(stdin.readline())
    for _ in range(cases):
        ds = DisjointSet()
        friendships = int(stdin.readline())
        for _ in range(friendships):
            friends = stdin.readline().split()
            a, b = friends[0], friends[1]
            try:
                ds.find(a)
            except KeyError:
                ds.addSet(a)
            try:
                ds.find(b)
            except:
                ds.addSet(b)
            ds.unite(a, b)
            print(len(ds.sets[ds.find(a)]))

if __name__ == "__main__":
    main()
